compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -442,21 +447,25 @@ static BOOL_TYS: TyS<'static> = TyS {
     kind: ty::Bool,
     flags: TypeFlags::empty(),
     outer_exclusive_binder: DebruijnIndex::from_usize(0),
+    stable_hash: Fingerprint::ZERO,
 };
 
 impl<'a, 'tcx> HashStable<StableHashingContext<'a>> for Ty<'tcx> {
     fn hash_stable(&self, hcx: &mut StableHashingContext<'a>, hasher: &mut StableHasher) {
         let TyS {
-            ref kind,
+            kind,
 
             // The other fields just provide fast access to information that is
             // also contained in `kind`, so no need to hash them.
             flags: _,
 
             outer_exclusive_binder: _,
+
+            stable_hash,
         } = self.0.0;
 
-        kind.hash_stable(hcx, hasher);
+        assert_ne!(*stable_hash, Fingerprint::ZERO, "{:#?}", kind);
+        stable_hash.hash_stable(hcx, hasher);
```

Comment:
cc @michaelwoerister @wesleywiser Instead of recursing through `TyKind`, this hashes a previously-computed `Fingerprint`, is that set up to do some kind of smart "hash mixing", or could it have adverse consequences? (in either performance or the quality of the resulting hash)

Reply:
We do the same thing in many places already. It's fine as long as the hash being hashed is wide enough and of good quality.